remoteconfig_test.go

```diff
@@ -569,42 +566,108 @@ func (s *RemoteConfigSuite) TestdownloadJSONValidate() {
+	assert.NotNil(s.T(), err)
+	assert.Regexp(s.T(), regexp.MustCompile("Request to 'http://\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}:\\d{1,5}' returned non-200 OK status '404: Not Found'"), err.Error())
+}
+
+func (s *RemoteConfigSuite) TestLoadConfigFromURLError() {
+	c := &SQSQueueConfig{}
+	err := LoadConfigFromURL("invalid", c)
+	assert.NotNil(s.T(), err)
+	assert.EqualError(s.T(), err, "Get invalid: unsupported protocol scheme \"\"")
+}
+
+func (s *RemoteConfigSuite) TestReadJSONValidate() {
+	configJSON := `
+	{
+		"sqs_client" : {
+			"region" : "us-east-1",
+			"endpoint" : "http://localhost:3000/sqs"
+		},
+		"sqs_queue" : {
+			"region" : "us-east-1",
+			"aws_account_id" : "345833302425",
+			"queue_name" : "testQueue"
+		},
+		"dynamodb_client" : {
+			"region" : "us-east-1",
+			"endpoint" : "http://localhost:8000/dynamodb"
+		},
+		"dynamodb_table" : {
+			"table_name" : "testTable"
+		},
+		"str" : "testStr",
+		"storage_config" : {
+			"provider" : "aws",
+			"location" : "us-west-2"
+		},
+		"storage_config_slice" : [{
+			"provider" : "aws",
+			"location" : "us-west-2"
+		},
+		{
+			"provider" : "aws",
+			"location" : "us-east-1"
+		}],
+		"storage_config_map": {
+			"one": {
+				"provider": "aws",
+				"location": "us-west-2"
+			}
+		},
+		"str_slice": [ "hello" ],
+		"map_str_str": { "key": "value" }
+	}`
+
+	cfgBuffer := bytes.NewBufferString(configJSON)
+
+	c := &SampleConfig{}
+	err := ReadJSONValidate(cfgBuffer, c)
+	assert.Nil(s.T(), err)
+}
+
+func (s *RemoteConfigSuite) TestReadJSONValidateInvalidJSON() {
+	cfgBuffer := bytes.NewBufferString("Not JSON")
+
 	c := &SampleConfig{}
-	err := DownloadJSONValidate(ts.URL, c)
+	err := ReadJSONValidate(cfgBuffer, c)
 	assert.NotNil(s.T(), err)
-	assert.Equal(s.T(), fmt.Errorf("Download of JSON failed, URL = %s, Response Code = %d", ts.URL, http.StatusNotFound), err)
+	assert.Equal(s.T(), errors.New("Failed to decode JSON, with error, invalid character 'N' looking for beginning of value"), err)
 }
 
-func (s *RemoteConfigSuite) TestdownloadJSONValidateErrorValidation() {
-	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		fmt.Fprintln(w, `{}`)
-	}))
-	defer ts.Close()
+func (s *RemoteConfigSuite) TestReadJSONValidateErrorValidation() {
+	cfgBuffer := bytes.NewBufferString("{}")
 
 	c := &SampleConfig{}
-	err := DownloadJSONValidate(ts.URL, c)
+	err := ReadJSONValidate(cfgBuffer, c)
 	assert.NotNil(s.T(), err)
 	assert.Equal(s.T(), errors.New("Field: SQSQueue, not set"), err)
 }
 
-func (s *RemoteConfigSuite) TestdownloadJSONValidateErrorInvalidJSON() {
+func (s *RemoteConfigSuite) TestReadJSONValidateErrorInvalidJSON() {
 	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		fmt.Fprintln(w, "This is NOT JSON")
 	}))
 	defer ts.Close()
 
+	resp, _ := http.Get(ts.URL)
+	defer resp.Body.Close()
+
 	c := &SampleConfig{}
-	err := DownloadJSONValidate(ts.URL, c)
+	err := ReadJSONValidate(resp.Body, c)
```

Comment:
This could either:

 1. Become an Example() instead of a test (its a great illustration of how to use this -- but maybe use valid JSON)
 2. Omit the test server / defer statement, and make this line here into: `err := ReadJSONValidate("NOT JSON", c)`